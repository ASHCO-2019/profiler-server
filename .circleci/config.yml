version: 2.1

executors:
  node:
    docker:
      - image: circleci/node:10.15

commands:
  checkout-and-dependencies:
    description: "Checkout and install dependencies, managing a cache"
    steps:
      - checkout
      - restore_cache:
          keys:
            - v1-dependencies-{{ checksum "package.json" }}-{{ checksum "yarn.lock" }}
            - v1-dependencies-{{ checksum "package.json" }}
            - v1-dependencies-
      # With --frozen-lockfile, the installation will fail if the lockfile is
      # outdated compared to package.json.
      - run: yarn install --frozen-lockfile
      - save_cache:
          paths:
            - node_modules
          key: v1-dependencies-{{ checksum "package.json" }}-{{ checksum "yarn.lock" }}

workflows:
  version: 2
  master:
    jobs:
      - tests
      - lint
      - build
      - flow
      - docker

jobs:
  lint:
    executor: node
    steps:
      - checkout-and-dependencies
      - run: yarn lint

  tests:
    executor: node
    steps:
      - checkout-and-dependencies
      - run: yarn test --runInBand

  flow:
    executor: node
    steps:
      - checkout-and-dependencies
      - run: yarn flow:ci

  build:
    executor: node
    steps:
      - checkout-and-dependencies
      - run: |
          yarn build:clean
          yarn build

  docker:
    executor: node
    steps:
      # This job doesn't need the cached node_modules because they'll be
      # redownloaded when building the docker image.
      - checkout
      # This sets up a remote environment that's necessary to run docker commands.
      - setup_remote_docker
      - run:
          name: "Build the Docker image"
          command: yarn docker:build --pull --build-arg circle_build_url=$CIRCLE_BUILD_URL
      - run:
          name: "Start the docker image in a container and check it's running"
          command: |
            yarn docker:run:detached
            # Wait up to 10 seconds that the server is launched.
            timeout 10s sh \<<'EOF'
            while ! docker exec profiler-server curl --silent --show-error --fail http://localhost:8000/__version__ ; do
              sleep 1
            done
            EOF
            yarn docker:stop
