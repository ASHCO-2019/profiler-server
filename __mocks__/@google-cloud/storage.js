// @flow
// This was mostly stolen from https://gist.github.com/nfarina/90ba99a5187113900c86289e67586aaa.
// And then rewritten to remove the dependency to stream-buffers and add the
// `exists` method.
// Thanks!

//
// Quick & Dirty Google Cloud Storage emulator for tests.
//
// `new MockStorage().bucket('my-bucket').file('my_file').createWriteStream()`
//

import { Readable } from 'stream';
import { Concatenator } from '../../src/utils/streams';

class MockStorage {
  static buckets: { [name: string]: MockBucket } = {};
  static cleanUp() {
    this.buckets = {};
  }

  bucket(name: string) {
    return (
      MockStorage.buckets[name] ||
      (MockStorage.buckets[name] = new MockBucket(name))
    );
  }
}

class MockBucket {
  name: string;
  files: { [path: string]: MockFile };

  constructor(name: string) {
    this.name = name;
    this.files = {};
  }

  file(path: string) {
    return this.files[path] || (this.files[path] = new MockFile(path));
  }
}

type Metadata = {
  +metadata: { ... },
  ...
};

class MockFile {
  name: string;
  contents: Buffer;
  metadata: Metadata;

  constructor(name: string) {
    this.name = name;
    this.contents = Buffer.alloc(0);
    this.metadata = {};
  }

  get() {
    return [this, this.metadata];
  }

  setMetadata(metadata: Metadata) {
    const customMetadata = { ...this.metadata.metadata, ...metadata.metadata };
    this.metadata = { ...this.metadata, ...metadata, metadata: customMetadata };
  }

  createReadStream() {
    const readable = new Readable({ autoDestroy: true });
    readable.push(this.contents);
    readable.push(null);
    return readable;
  }

  createWriteStream({ metadata }: Object) {
    this.setMetadata(metadata);
    const writable = new Concatenator();
    writable.on('finish', () => {
      this.contents = writable.transferContents();
      writable.destroy();
    });
    return writable;
  }
}

export { MockStorage as Storage, MockBucket as Bucket, MockFile as File };
