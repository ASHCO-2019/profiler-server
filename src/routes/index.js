/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import Router from '@koa/router';
import cors from '@koa/cors';

import { dockerFlowRoutes } from './dockerflow';
import { publishRoutes } from './publish';
import { cspReportRoutes } from './cspReport';

import { versioning } from '../middlewares';

export function routes() {
  const router = new Router();
  router.use(dockerFlowRoutes().routes());
  router.use(cspReportRoutes().routes());

  // Versioning and CORS applies only to API routes, that's why we specify it here.
  router.use(versioning(1));
  // We use the default configuration for cors, that is we allow all origins and
  // all methods.
  router.use(cors());
  router.use(publishRoutes().routes());
  return router;
}
