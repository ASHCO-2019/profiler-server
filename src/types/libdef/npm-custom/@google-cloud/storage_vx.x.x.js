/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

// In this file we type only the API we use in the project. If we use more API
// later this will need to be completed.

declare module '@google-cloud/storage' {
  // Source: https://googleapis.dev/nodejs/storage/latest/global.html#CreateWriteStreamOptions
  declare type CreateWriteStreamOptions = {|
    configPath: string,
    contentType: string,
    gzip: boolean | 'auto',
    metadata: { ... },
    offset: number,
    predefinedAcl: string,
    private: boolean,
    public: boolean,
    resumable: boolean,
    uri: string,
    userProject: string,
    validation: 'md5' | 'crc32c' | false,
  |};
  // Source: https://googleapis.dev/nodejs/storage/latest/File.html
  declare export class File {
    constructor(
      bucket: Bucket,
      name: string,
      options?: $Shape<{|
        encryptionKey: string,
        generation: number,
        kmsKeyName: string,
        userProject: string,
      |}>
    ): File;

    // Source: https://googleapis.dev/nodejs/storage/latest/File.html#createWriteStream
    createWriteStream(
      options?: $Shape<CreateWriteStreamOptions>
    ): stream$Writable;
  }

  // Source: https://googleapis.dev/nodejs/storage/latest/Bucket.html
  declare export class Bucket {
    name: string;

    constructor(
      storage: Storage,
      name: string,
      options?: $Shape<{| userProject: string |}>
    ): Bucket;

    // Source: https://googleapis.dev/nodejs/storage/latest/Bucket.html#exists
    exists(
      options?: $Shape<{| userProject: string |}>,
      callback?: (err: ?Error, exists: boolean) => mixed
    ): Promise<[boolean]>;

    // Source: https://googleapis.dev/nodejs/storage/latest/Bucket.html#file
    file(
      name: string,
      options?: $Shape<{|
        generation: string | number,
        encryptionKey: string,
        kmsKeyName: string,
      |}>
    ): File;
  }

  // Source: https://googleapis.dev/nodejs/storage/latest/global.html#StorageOptions
  declare type StorageOptions = {|
    projectId: string,
    keyFilename: string,
    email: string,
    credentials: $Shape<{|
      client_email: string,
      private_key: string,
    |}>,
    autoRetry: boolean,
    maxRetries: number,
    promise: typeof Promise,
  |};

  // Source: https://googleapis.dev/nodejs/storage/latest/Storage.html
  declare export class Storage {
    constructor(options: ?$Shape<StorageOptions>): Storage;

    // Source: https://googleapis.dev/nodejs/storage/latest/Storage.html#bucket
    bucket(
      name: string,
      options?: $Shape<{| kmsKeyName: string, userProject: string |}>
    ): Bucket;
  }
}
