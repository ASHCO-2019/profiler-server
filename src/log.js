/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
//
// Shamelessly stolen from Firefox Send.

import mozlog from 'mozlog';

import { config } from './config';

const isProduction = config.env === 'production';

export const getLogger = mozlog({
  app: 'FirefoxProfiler',
  level: isProduction ? 'INFO' : 'verbose',
  fmt: isProduction ? 'heka' : 'pretty',
  debug: !isProduction, // This asserts a correct usage of the library
});

type ExtractReturnType = <T>((...rest: any) => T) => T;
export type Logger = $Call<ExtractReturnType, typeof getLogger>;
