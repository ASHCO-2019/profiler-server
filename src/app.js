/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import Koa from 'koa';

import routes from './routes';

export function createApp() {
  const app = new Koa();

  // Adding the main endpoints for this app.
  // koa-router exposes 2 middlewares:
  // - the result of routes() returns the configured routes.
  // - the result of allowedMethods() configures the OPTIONS verb.
  const router = routes();
  app.use(router.routes());
  app.use(router.allowedMethods());

  return app;
}
