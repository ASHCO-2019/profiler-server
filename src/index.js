/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import Koa from 'koa';

import config from './config';
import routes from './routes';

const app = new Koa();

const router = routes();
app.use(router.routes());
app.use(router.allowedMethods());

app.listen(config.httpPort);
console.log('The HTTP server is listening on port %s.', config.httpPort);
