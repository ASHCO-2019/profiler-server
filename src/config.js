/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import convict from 'convict';
import { getLogger } from './log';

const log = getLogger('config');

const conf = convict({
  env: {
    doc: 'The application environment.',
    format: ['production', 'development', 'test'],
    default: 'development',
    env: 'NODE_ENV',
  },
  httpPort: {
    format: 'port',
    default: 4243,
    // The environment variable's name is required by Dockerflow.
    // see https://github.com/mozilla-services/Dockerflow#containerized-app-requirements
    env: 'PORT',
  },
});

if (conf.get('env') !== 'test') {
  // We don't want to run a local configuration file for tests so that we
  // ensure that they'll always run the same in all environments.
  try {
    // Load local configuration if present.
    conf.loadFile('./local-config.json');
    log.debug(
      'local_configuration',
      `Local configuration file 'local-config.json' was found and loaded.`
    );
  } catch (e) {
    // But it's OK if it's absent.
    log.debug(
      'local_configuration',
      `Local configuration file 'local-config.json' was not found, but it's OK.`
    );
  }
}

conf.validate();

type Config = {|
  +env: string,
  +httpPort: number,
|};

export const config: Config = conf.getProperties();
log.info(
  'configuration_success',
  'Configuration info has been loaded successfully.'
);
