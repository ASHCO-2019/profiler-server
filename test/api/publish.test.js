/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow
import supertest from 'supertest';

import crypto from 'crypto';

import { Storage as MockStorage } from '../../__mocks__/@google-cloud/storage';

import { config } from '../../src/config';
import { createApp } from '../../src/app';

beforeEach(() => MockStorage.cleanUp());
afterEach(() => MockStorage.cleanUp());

describe('publishing endpoints', () => {
  function setup() {
    const agent = supertest(createApp().callback());
    return agent.post('/compressed-store').type('text');
  }

  it('uploads data all the way to google storage when using a content length', async () => {
    const content = 'aaaa';
    const req = setup();
    await req.send(content).expect(200);

    const contentHash = crypto
      .createHash('sha1')
      .update(content)
      .digest('hex');

    expect(MockStorage.buckets).toHaveProperty(config.gcsBucket);

    const bucket = MockStorage.buckets[config.gcsBucket];
    expect(bucket.files).toHaveProperty(contentHash);

    const file = bucket.files[contentHash];
    expect(file).toHaveProperty('contents', Buffer.from(content));
    expect(file).toHaveProperty('name', contentHash);
    expect(file).toHaveProperty(
      'metadata',
      expect.objectContaining({
        cacheControl: 'max-age: 365000000, immutable',
        contentEncoding: 'gzip',
        contentType: 'text/plain',
      })
    );
  });

  it('uploads data all the way to google storage when using chunked encoding', async () => {
    const content = 'aaaa';
    const req = setup();

    // When using the low-level API "write", Node generates a "chunked encoding"
    // request without a Content-Length. This is exactly what we want to check
    // here.
    // Unfortunately there's no easy way to assert we're in this node, we'll
    // have to rely on the heuristic.
    req.write(content);

    await req.expect(200);

    const contentHash = crypto
      .createHash('sha1')
      .update(content)
      .digest('hex');

    expect(MockStorage.buckets).toHaveProperty(config.gcsBucket);
    const bucket = MockStorage.buckets[config.gcsBucket];
    expect(bucket.files).toHaveProperty(contentHash);
    const file = bucket.files[contentHash];
    expect(file).toHaveProperty('contents', Buffer.from(content));
    expect(file).toHaveProperty('name', contentHash);
    expect(file).toHaveProperty(
      'metadata',
      expect.objectContaining({
        cacheControl: 'max-age: 365000000, immutable',
        contentEncoding: 'gzip',
        contentType: 'text/plain',
      })
    );
  });

  it('returns an error when the length is too big', async () => {
    const req = setup();
    await req.set('Content-Length', String(1024 * 1024 * 1024)).expect(413);
  });

  it('returns an error when the sent data is bigger than the length', async () => {
    jest.spyOn(process.stdout, 'write').mockImplementation(() => {});
    const req = setup();
    await req
      .set('Content-Length', String(3))
      .send('aaaa')
      .expect(400);
    expect(process.stdout.write).toHaveBeenCalledWith(
      expect.stringContaining('server_error')
    );
  });

  it('returns an error when the pushed buffer is too big', async () => {
    jest.spyOn(process.stdout, 'write').mockImplementation(() => {});
    const req = setup();

    // When using the low-level API "write", Node generates a "chunked encoding"
    // request without a Content-Length. This is exactly what we want to check
    // here.
    await req.write(Buffer.alloc(33 * 1024 * 1024));

    const error = await req.then(
      () => {
        throw new Error(`The request shouldn't succeed.`);
      },
      e => e
    );

    // We don't get a 413 because the connection is reset without a response.
    expect(error.code).toBe('ECONNRESET');
    // This check asserts that we get the error using the
    // LengthCheckerPassThrough transform. This asserts that we stop the stream
    // early even without a Content-Length header.
    expect(process.stdout.write).toHaveBeenCalledWith(
      expect.stringContaining('length-checker-length-error')
    );
    expect(process.stdout.write).toHaveBeenCalledWith(
      expect.stringContaining('server_error')
    );
  });
});
