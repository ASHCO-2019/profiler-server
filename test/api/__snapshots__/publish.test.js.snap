// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API versioning returns an error when an \`accept\` header is specified with an unsupported version 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1mFirefoxProfiler.app[22m.[1m[31mERROR[39m[22m: server_error {\\"error\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\",\\"stack\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\\\n    at Object.throw (/home/julien/travail/git/profiler-server/node_modules/koa/lib/context.js:97:11)\\\\n    at /home/julien/travail/git/profiler-server/src/middlewares/versioning.js:50:7\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at next (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:45:18)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:346:16\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:36:12\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:351:31)\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:42:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:34:12\\\\n    at Application.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:162:12)\\\\n    at Server.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:144:19)\\\\n    at Server.emit (events.js:210:5)\\\\n    at parserOnIncoming (_http_server.js:745:12)\\\\n    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:17)\\"}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`API versioning returns an error when an accept header is specified with several unacceptable values 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1mFirefoxProfiler.app[22m.[1m[31mERROR[39m[22m: server_error {\\"error\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\",\\"stack\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\\\n    at Object.throw (/home/julien/travail/git/profiler-server/node_modules/koa/lib/context.js:97:11)\\\\n    at /home/julien/travail/git/profiler-server/src/middlewares/versioning.js:50:7\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at next (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:45:18)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:346:16\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:36:12\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:351:31)\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:42:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:34:12\\\\n    at Application.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:162:12)\\\\n    at Server.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:144:19)\\\\n    at Server.emit (events.js:210:5)\\\\n    at parserOnIncoming (_http_server.js:745:12)\\\\n    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:17)\\"}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`API versioning returns an error when an accept header is specified without a version at all 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1mFirefoxProfiler.app[22m.[1m[31mERROR[39m[22m: server_error {\\"error\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\",\\"stack\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\\\n    at Object.throw (/home/julien/travail/git/profiler-server/node_modules/koa/lib/context.js:97:11)\\\\n    at /home/julien/travail/git/profiler-server/src/middlewares/versioning.js:50:7\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at next (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:45:18)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:346:16\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:36:12\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:351:31)\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:42:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:34:12\\\\n    at Application.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:162:12)\\\\n    at Server.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:144:19)\\\\n    at Server.emit (events.js:210:5)\\\\n    at parserOnIncoming (_http_server.js:745:12)\\\\n    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:17)\\"}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`API versioning returns an error when an invalid \`accept\` header is specified 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1mFirefoxProfiler.app[22m.[1m[31mERROR[39m[22m: server_error {\\"error\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\",\\"stack\\":\\"NotAcceptableError: The header 'Accept' should have the value application/vnd.firefox-profiler+json; version=1\\\\n    at Object.throw (/home/julien/travail/git/profiler-server/node_modules/koa/lib/context.js:97:11)\\\\n    at /home/julien/travail/git/profiler-server/src/middlewares/versioning.js:50:7\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at next (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:45:18)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:346:16\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:36:12\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:351:31)\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:42:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:34:12\\\\n    at Application.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:162:12)\\\\n    at Server.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:144:19)\\\\n    at Server.emit (events.js:210:5)\\\\n    at parserOnIncoming (_http_server.js:745:12)\\\\n    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:17)\\"}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`API versioning returns an error when no \`accept\` header is specified 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[1mFirefoxProfiler.app[22m.[1m[31mERROR[39m[22m: server_error {\\"error\\":\\"BadRequestError: The header 'Accept' is missing.\\",\\"stack\\":\\"BadRequestError: The header 'Accept' is missing.\\\\n    at Object.throw (/home/julien/travail/git/profiler-server/node_modules/koa/lib/context.js:97:11)\\\\n    at /home/julien/travail/git/profiler-server/src/middlewares/versioning.js:25:7\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at next (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:45:18)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:346:16\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:44:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa-compose/index.js:36:12\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/@koa/router/lib/router.js:351:31)\\\\n    at dispatch (/home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:42:32)\\\\n    at /home/julien/travail/git/profiler-server/node_modules/koa/node_modules/koa-compose/index.js:34:12\\\\n    at Application.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:162:12)\\\\n    at Server.handleRequest (/home/julien/travail/git/profiler-server/node_modules/koa/lib/application.js:144:19)\\\\n    at Server.emit (events.js:210:5)\\\\n    at parserOnIncoming (_http_server.js:745:12)\\\\n    at HTTPParser.parserOnHeadersComplete (_http_common.js:115:17)\\"}
",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
